Clean up DNS e2e wrt Jessie/Wheezy

thockin opened the pull request:
The distinction between Jessie and Wheezy dates back to https://github.com/kubernetes/kubernetes/issues/10161 (mid 2015).  For this test, wheezy has long ago been replaced by agnhost.  Now Jessie is gone, too.

/kind cleanup
Fixes  #126936

```release-note
NONE
```

thockin committed:
Clean up DNS e2e wrt Jessie/Wheezy

skitt commented on test/e2e/network/dns.go:
```diff
@@ -79,15 +77,13 @@ var _ = common.SIGDescribe("DNS", func() {
-		pod := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd, dnsTestPodHostName, dnsTestServiceName)
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		pod := createDNSPod(f.Namespace.Name, probeCmd, dnsTestPodHostName, dnsTestServiceName)
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -57,15 +57,13 @@ var _ = common.SIGDescribe("DNS", func() {
-		pod := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd, dnsTestPodHostName, dnsTestServiceName)
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		pod := createDNSPod(f.Namespace.Name, probeCmd, dnsTestPodHostName, dnsTestServiceName)
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -100,15 +96,13 @@ var _ = common.SIGDescribe("DNS", func() {
-		pod := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd, dnsTestPodHostName, dnsTestServiceName)
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		pod := createDNSPod(f.Namespace.Name, probeCmd, dnsTestPodHostName, dnsTestServiceName)
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -120,15 +114,13 @@ var _ = common.SIGDescribe("DNS", func() {
-		pod := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd, dnsTestPodHostName, dnsTestServiceName)
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		pod := createDNSPod(f.Namespace.Name, probeCmd, dnsTestPodHostName, dnsTestServiceName)
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -171,17 +163,15 @@ var _ = common.SIGDescribe("DNS", func() {
 		pod.ObjectMeta.Labels = testServiceSelector
 
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -224,17 +214,15 @@ var _ = common.SIGDescribe("DNS", func() {
 		pod.ObjectMeta.Labels = testServiceSelector
 
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -264,19 +252,17 @@ var _ = common.SIGDescribe("DNS", func() {
 		pod1.Spec.Subdomain = serviceName
 
-		validateDNSResults(ctx, f, pod1, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod1, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod1, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -307,19 +293,17 @@ var _ = common.SIGDescribe("DNS", func() {
 		pod1.Spec.Subdomain = serviceName
 
-		validateDNSResults(ctx, f, pod1, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod1, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod1, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -595,7 +571,7 @@ var _ = common.SIGDescribe("DNS", func() {
 			},
 		}
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented on test/e2e/network/dns.go:
```diff
@@ -630,18 +606,16 @@ var _ = common.SIGDescribe("DNS HostNetwork", func() {
 		pod.Spec.HostNetwork = true
 		pod.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
-		validateDNSResults(ctx, f, pod, append(wheezyFileNames, jessieFileNames...))
+		validateDNSResults(ctx, f, pod, append(fileNames))
```
```suggestion
		validateDNSResults(ctx, f, pod, fileNames)
```

skitt commented:
/lgtm

skitt commented:
/hold to give @aojea a chance to review

skitt commented on test/e2e/network/dns_common.go:
```diff
@@ -373,17 +372,9 @@ func createDNSPod(namespace, wheezyProbeCmd, jessieProbeCmd, podHostName, servic
-		VolumeMounts: mounts,
-	}
-
-	dnsPod.Spec.Containers = append(dnsPod.Spec.Containers, querier, jessieQuerier)
```
This shouldn’t have been removed entirely, should it?

```
dnsPod.Spec.Containers = append(dnsPod.Spec.Containers, querier)
```

thockin replied:
yep, found that, too. :)

thockin commented:
/retest

aojea commented on test/e2e/network/dns_common.go:
```diff
@@ -373,17 +372,10 @@ func createDNSPod(namespace, wheezyProbeCmd, jessieProbeCmd, podHostName, servic
+	querier := e2epod.NewAgnhostContainer("querier", mounts, nil, probeCmd)
 	querier.Command = []string{"sh", "-c"}
 
-	jessieQuerier := v1.Container{
```
the problem I see here is that we are loosing coverage with the libc implementation, right now we have `muslc` and `libc` covered.

We should not completely remove it and instead use one image that has debian or other libc implementation

thockin replied:
You're not wrong.  It looks like agnhost is alpine:

```
ldd `which dig`
	/lib/ld-musl-x86_64.so.1 (0x7fa21cb68000)
	libdns-9.16.27.so => /usr/lib/libdns-9.16.27.so (0x7fa21c975000)
        ...
```

Thast said, if we establish an image just for DNS testing which is libc based, someone has to perpetually build new versions of it.

I wonder if we can talk $someone into building agnhost on both alpine and debian bases?

aojea replied:
just replace s/jessie/debian/ for names and s/debian:jessie/debian:stable-slim/ for images

aojea replied:
let me do the building of the image

danwinship replied:
> we are losing coverage with the libc implementation

Haha, no, we lost the relevant coverage _years_ ago. (But it doesn't matter.)

The timeline here is:
1. Wheezy has a broken DNS implementation which gives ndots-like behavior even if you don't specify ndots. (I think? That's what it seems like from the old discussion.)
2. Kubernetes pre-1.0 assumed that behavior.
3. Jessie fixed that behavior, breaking Kubernetes (#10161).
4. Kubelet was fixed to set `ndots:5` in resolv.conf, fixing Kubernetes under jessie. (#10266)
5. At this time, the e2e test images were all still based on wheezy, so the DNS e2e tests were modified to also test against special jessie-based images (since the wheezy-based tests would have passed with or without the `ndots` fix). It's not clear why it tested against both rather than _only_ testing against jessie, but [Tim commented](https://github.com/kubernetes/kubernetes/pull/10266#issuecomment-115103614) "We can reduce the test case back to one when we do a new dnsutils container" so it was clearly intended to be temporary.
6. ... but apparently everyone forgot about that, and mistakenly assumed that the point of the jessie image was that we wanted to test against jessie specifically, rather than that we had just needed something newer than wheezy.

Anyway, while it may happen to be the case that the current version of this test tests two different libcs, that is not the _intent_ of the case (and I don't think there is any reason to believe that libc and musl behave differently here anyway).

thockin replied:
If we build agnhost for alpine and debian, I think it would be OK to test both - these tests are fast and cheap, and musl HAS been a problem in the past (not wrt ndots but other things) and we only figured it out when other people reported bugs.

I don't love the idea of having a test image whose sole purpose is DNS, because we are expected to support it (the original bug behind this PR)

aojea replied:
> Haha, no, we lost the relevant coverage years ago. (But it doesn't matter.)

@danwinship jessie image is using libc, isn't it?

>  believe that libc and musl behave differently here anyway).

not a lot, but there is precedent https://github.com/kubernetes/kubernetes/issues/112135

danwinship replied:
OK... so, the syntax and semantics of `/etc/resolv.conf` are not sufficiently well defined that we can assume that all software in all containers will interpret it identically. So we need to have an e2e test for "kubelet writes out a resolv.conf that is compatible with the resolver libraries of languages/runtimes we expect people to use in pods". And the easiest way to test that something is interpreting resolv.conf correctly is to just have it resolve things, and the only thing we know for sure that pods will be able to resolve is the Kubernetes DNS stuff. So here we are.

But we don't need to test multiple resolvers in _every_ DNS test. We should split out the alternative-runtime DNS tests into a single separate test case, to make it easier to distinguish "your cluster DNS is broken" failures from "kubelet's resolv.conf generation doesn't work with some languages/runtimes" failures.

And maybe we should test more resolv.conf options, to be able to catch things like #112135?

aojea replied:
right know, by chance, we have this libc/muslc more or less covered .. not a lot but at least the basic is covered 

> And maybe we should test more resolv.conf options, to be able to catch things like https://github.com/kubernetes/kubernetes/issues/112135?

yeah, we can do a way more for testing, specially on e2e ... but it requires time  ... I've tried to use e2e testing for providing opportunities for people to contribute, but if people is not able to progress alone with minimum help then it steals time from the ones that don't have time ... and the ones that get their code merged leave once they got their reward of having a commit in the repo

aojea commented:
@thockin you are going to hate me https://github.com/kubernetes/kubernetes/pull/127046#discussion_r1743598169 :)

aojea commented on test/images/jessie-dnsutils/BASEIMAGE:
```diff
@@ -1,7 +0,0 @@
-linux/amd64=debian:jessie
```
linux/amd64=debian:stable-slim

aojea replied:
here https://github.com/kubernetes/kubernetes/pull/127129

aojea commented:
, the existing images used for test are outdated on most places, looking for alpine
```
grep -r alpine test/images/ | grep BASE | grep amd
test/images/nonewprivs/BASEIMAGE:linux/amd64=alpine:3.6
test/images/sample-apiserver/BASEIMAGE:linux/amd64=alpine:3.8
test/images/nginx-new/BASEIMAGE:linux/amd64=nginx:1.15-alpine
test/images/redis/BASEIMAGE:linux/amd64=alpine:3.6
test/images/agnhost/BASEIMAGE:linux/amd64=alpine:3.19
test/images/nginx/BASEIMAGE:linux/amd64=nginx:1.14-alpine
test/images/httpd/BASEIMAGE:linux/amd64=httpd:2.4.38-alpine
test/images/sample-device-plugin/BASEIMAGE:linux/amd64=alpine:3.8
test/images/httpd-new/BASEIMAGE:linux/amd64=httpd:2.4.39-alpine
test/images/ipc-utils/BASEIMAGE:linux/amd64=alpine:3.6
test/images/apparmor-loader/BASEIMAGE:linux/amd64=alpine:3.14
```

so I don't think that having jessie is a bigger problem than having alpine:3.6 that was EOL on 2019 https://alpinelinux.org/releases/

thockin commented:
so should we just leave it as Jessie because we don't have time to fix it "for real" and "for everyone" ?

As I said elsewhere, I am not really excited about adding yet another image which will inevitably become out of date.  If we make agnhost build for both debian and alpine, then we could get full coverage on both of the major variants.  I hear Dan's point that we don't need _FULL_ coverage, but these tests are pretty fast, and we could later factor it so that only critical ones run on both.

I'm happy to close this and file a TODO issue instead.  I doubt we'll ever do it unless someone steps up.  This change was attractive because it was easy, but I don't have much time to do it "right", at least not right now.

danwinship commented:
OK, apparently at one point OpenShift ran into a problem with clients that use the golang resolver too (https://github.com/openshift/cluster-dns-operator/pull/276)

BenTheElder commented:
> OK, apparently at one point OpenShift ran into a problem with clients that use the golang resolver too

That makes sense, the behavior is glibc-ish, but reimplemented.

Also not relevant for e2e test, but that behavior has changed before on some platforms (macOS) to move to the platform resolver even when not using cgo: https://danp.net/posts/macos-dns-change-in-go-1-20/

thockin commented:
I again probe for opinions on what folks want to do here.

I will not have time to turn this into a more robust test, using multiple resolver implementations, any time soon.  So it's either:
1) abandon it and leave it as-is
2) abandon it and just rename stuff
3) merge it
4) someone with more time picks it up

thockin committed:
Revert "Clean up DNS e2e wrt Jessie/Wheezy"

This reverts commit eeb36da2b97e87b3534c7bece9355cefc6f16472.

thockin commented:
We could convert this to a list-of-cases so it becomes easy to add more (e.g. Go's resolver) but it's going to need a bunch of refactoring.  "agnhost" is treated somewhat specially, so we probably want to do something like:

```
cases := []struct {
    name string
    mode dnsTestMode
    args dnsTestArgs
}{{
    name: "agnhost",
    mode: AgnHost,
}, {
    name: "jessie",
    mode: Image
    image: imageutils.GetE2EImage(imageutils.JessieDnsutils)
}}
for _, case := range cases {
    cmd, ctr, files := makeDNSTester(case.name, case.mode, namesToResolve, nil, "", f.Namespace.Name, framework.TestContext.ClusterDNSDomain, framework.TestContext.ClusterIsIPv6())
    // log cmd
    // accumulate ctr and pass them to createDNSPod()
    // accumulate files and pass them to validateDNSResults()
}
```

danwinship commented:
If you have time to do "abandon it and just rename stuff", I think that would be a big improvement over abandoning it completely. The current code is pretty confusing.

aojea commented:
/test pull-kubernetes-e2e-gce-storage-slow 

unrelated failures, it should not run also in this PR afaik

aojea commented:
@thockin you have severl problems because some function signature change

> # k8s.io/kubernetes/test/e2e/network
> test/e2e/network/dns.go:603:93: too many arguments in call to createDNSPod

there may more related , check the `typecheck` and `verify` jobs , they show more meaninful error messages

thockin committed:
Rename DNS tests and restructure

Get rid if "wheezy" (it's not wheezy) and set it up to process an
arbitrary list of probers, but stop short of adding more.

aojea commented:
/lgtm
